crates/risc0-bonsai/src/host.rs

```diff
@@ -13,278 +11,41 @@ use risc0_zkvm::{
-                    Err(e) => {
-                        error!(?e, "Failed to connect to Bonsai");
-                        std::thread::sleep(Duration::from_secs(5));
-                        continue 'client;
-                    }
-                };
-                'queue: loop {
-                    let request = if let Some(last_request) = last_request.clone() {
-                        debug!("Retrying last request after reconnection");
-                        last_request
-                    } else {
-                        trace!("Waiting for a new request");
-                        let req: BonsaiRequest = rx.recv().expect("bonsai client sender is dead");
-                        // Save request for retries
-                        last_request = Some(req.clone());
-                        req
-                    };
-                    match request {
-                        BonsaiRequest::UploadImg {
-                            image_id,
-                            buf,
-                            notify,
-                        } => {
-                            debug!(%image_id, "Bonsai:upload_img");
-                            let res = client.upload_img(&image_id, buf);
-                            let res = unwrap_bonsai_response!(res, 'client, 'queue);
-                            let _ = notify.send(res);
-                        }
-                        BonsaiRequest::UploadInput { buf, notify } => {
-                            debug!("Bonsai:upload_input");
-                            let res = client.upload_input(buf);
-                            let res = unwrap_bonsai_response!(res, 'client, 'queue);
-                            let _ = notify.send(res);
-                        }
-                        BonsaiRequest::Download { url, notify } => {
-                            debug!(%url, "Bonsai:download");
-                            let res = client.download(&url);
-                            let res = unwrap_bonsai_response!(res, 'client, 'queue);
-                            let _ = notify.send(res);
-                        }
-                        BonsaiRequest::CreateSession {
-                            img_id,
-                            input_id,
-                            assumptions,
-                            notify,
-                        } => {
-                            debug!(%img_id, %input_id, "Bonsai:create_session");
-                            // TODO: think about whether we should have a case where we use Bonsai with only execute mode
-                            let res = client.create_session(img_id, input_id, assumptions, false);
-                            let res = unwrap_bonsai_response!(res, 'client, 'queue);
-                            let _ = notify.send(res);
+use tracing::{error, info, warn};
+
+macro_rules! unwrap_bonsai_response {
+    ($bonsai_call:expr) => (
+        loop {
+            let response = $bonsai_call;
```

Comment:
Instead of `loop`ing forever until a good response is received... why don't we use backoff crate to retry some number of times before failing?

Reply:
let's say we use backoff (I agree that's best practice), what happens when we fail due to some connection problems etc.? what should the prover do? I thought it will just retry proving --- which is nearly the same as just retrying indefinitely

Reply:
Right... but backoff retry:
1. Simplifies / potentially removes the need for this macro.
2. Retries in a better way by increasing the time between retries so that we have higher potential of succeeding at the next retry than when retrying every 5 seconds or something like that.